sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala

```diff
@@ -68,6 +70,14 @@ class SimpleFunctionRegistry extends FunctionRegistry {
   override def lookupFunction(name: String): Option[ExpressionInfo] = {
     functionBuilders.get(name).map(_._1)
   }
+
+  override def copy(): SimpleFunctionRegistry = {
```

Comment:
OK this is technically not part of this, but since you are modifying it ... shouldn't all of these catalogs be thread safe? Since they can be called by multiple threads.